chore(appium): experimental build via tsc

boneskull opened the pull request:
See #17807.

This shows that `appium` can be built via `tsc`.

Here's [a gist which shows the output difference](https://gist.github.com/boneskull/a0d3b3659a32830335ed7b9f87bfe31e)

boneskull committed:
chore(appium): experimental build via tsc

jlipps commented on config/tsconfig.base.json:
```diff
@@ -7,12 +7,12 @@
     "strictNullChecks": true,
     "removeComments": false,
-    "target": "es2015",
+    "target": "ESNext",
```
don't we want to make sure the target matches our lowest version of node?

boneskull replied:
hmm, yeah, probably.  I'm not sure what that is.  I'll have to figure it out.

boneskull replied:
OK, this is ES2020 for Node.js v14.17.0 as per [the kangax table](https://kangax.github.io/compat-table/es2016plus/)

boneskull committed:
fix(build): set target to ES2020
As per https://kangax.github.io/compat-table/es2016plus/ our soon-to-be-target lowest platform (Node.js v14.17.0) supports ES2020 in full.  Except tail call optimization, because reasons

boneskull commented:
If merged, this will close #17746

boneskull commented:
@jlipps @mykola-mokhnach @KazuCocoa 

This looks pretty good to me.  Should I expand this to compile the whole monorepo?

KazuCocoa reviewed (approved):
oh, looks cool to me, 🚀

boneskull committed:
chore: use tsc for compilation across the board

- `@appium/gulp-plugins` does not use `tsc` and continues to use...whatever it uses.
- `@appium/doctor` now has generated declarations, though I didn't bother to do anything with the types
- rewrote swaths of build scripts:
  - removed the `fix` and `lint` scripts from each workspace (package) since I don't think they get used and they are of limited value
  - for tests and autoinstallation of extensions, added `ts-node` for on-the-fly compilation
  - workspaces _can_ define their own `build` and `clean` scripts (both should be required if either is used).  these do not currently get run when running `npm run dev`, but they do occur on a bare `npm install` from the monorepo root or a `npm run reinstall`.
  - `npm run rebuild` does a fresh rebuild, but does not clean any `node_modules` dirs nor does it reinstall anything.
  - removed `prepublishOnly` since `preversion` does the same thing
  - `npm test` now runs `build`/`lint` in parallel
- postinstall script of `appium` ("autoinstall") script now must call `npm run build` if in a fresh dev environment. this is because a) lifecycle scripts of packages run before lifecycle scripts in the monorepo root, and b) `postinstall` runs before `prepare`. so there's really no way around it; even using `ts-node` fails because other modules depend on `@appium/support/index.js` which expects the pkg to be built.

Closes #17746
Closes #17807

boneskull committed:
chore: use foreground-scripts=true in .npmrc

this will give more visibility into what is happening during install

boneskull committed:
chore: increase default test timeout in CI

boneskull committed:
chore(ci): remove old script from CI

boneskull committed:
fix(fake-driver): copy screen.png on build

boneskull committed:
chore(appium,docutils,ci): remove doc gen from CI

What used to be the `generate-docs` script no longer works, since the `commands-yml` directory no longer exists.

Unclear to me what should be done here at this point, but it makes the "build command documentation" job pointless.

jlipps reviewed (approved):
awesome! lgtm once CI is happy. and did you manually run scripts that aren't run in CI just to double check they still work?

boneskull commented on .npmrc:
```diff
@@ -1,4 +1,4 @@
 package-lock=true
 save-exact=true
-
+foreground-scripts=true
```
it's new-ish. they added it because people were begging in `postinstall` scripts.

boneskull commented:
yeah I don't think this will work but it's a nice thought

```bash
for $script in $(cat package.json | '.scripts | keys[]'); do 
  npm run $script
done`
```

boneskull commented:
@jlipps btw the only reason CI is failing is because stuff that was silently failing no longer does so silently.

boneskull commented:
@jlipps OK, so I removed the "build documentation" job from CI because it hasn't been working in quite some time and it was failing silently.  Historically, it was trying to run the `parse-yml-commands` script, but that script is looking for files that no longer exist (`commands-yml/`, `template.md`).
